### PR TITLE
Guard missing Spark Job Server geoprocessing context

### PR DESCRIPTION
**Pull request by kellyi**
This PR addresses an error whereby a job submitted to Spark Job Server would occasionally fail because it could not find the "geoprocessing" context. The changes here add a check to `sjs_submit` to see whether a post request to SJS has failed with a message of 'context geoprocessing not found'. If so, we now make a put request to reset the geoprocessing service's contexts and -- if it's successful -- we try re-submitting the job.

**Testing**
- get this branch, `vagrant up`, then use httpie or curl to delete the geoprocessing context from SJS:

```sh
http DELETE :8090/contexts/geoprocessing
```

- check the service's contexts by making a get request to `:8090/contexts/` and verify that it's an empty array

- add this print statement at line 76 of `src/mmw/apps/modeling/geoprocessing.py`:

```py
print('error => {}'.format(error))
```
- run `./scripts/debugcelery.sh`
- open `localhost:8000` in the browser and select a 1 sq km area of interest
- switch back to the `debugcelery.sh` window and confirm that you see the print statement reporting the error but that the job continues and completes instead of throwing an Exception
- once the job's complete, check the frontend to see that results have been returned
- delete the geoprocessing context again with httpie or curl:

```sh
http DELETE :8090/contexts/geoprocessing
```

- run a MapShed job on your AOI, verify that you see the print statement above again in `debugcelery`, and verify that the job eventually completes and doesn't create an error

Connects #1476

### CODE REVIEW

File: src/mmw/apps/modeling/geoprocessing.py [hunk 1]
```diff
@@ -71,6 +72,16 @@ def sjs_submit(host, port, args, data, retry=None):
             except MaxRetriesExceededError:
                 raise Exception('No slots available in Spark JobServer.\n'
                                 'Details = {}'.format(response.text))
+        elif error['result'] == 'context geoprocessing not found':
+            geop_context_url = '{}/contexts?reset=reboot'.format(base_url)
+            context_response = requests.put(geop_context_url)
```
Comment (kellyi) [inline src/mmw/apps/modeling/geoprocessing.py]: An alternate possibility here would be to send this to Spark Job Server to create the "geoprocessing" context:

```py
geop_context_url = '{}/contexts/geoprocessing'.format(base_url)
context_response = requests.post(geop_context_url, data='')
```

Instead of resetting the contexts back to the initial config, this would create the "geoprocessing" context.

Comment (rajadain) [inline src/mmw/apps/modeling/geoprocessing.py]: Rebooting might be safer, because we're reusing the default context definition, so if we ever change its name we won't have to change it in multiple places. This also protects against other errors that could have happened: it is generally cleaner to reboot the whole thing than to try and piece it back together.

File: src/mmw/apps/modeling/geoprocessing.py [hunk 2]
```diff
@@ -71,6 +72,16 @@ def sjs_submit(host, port, args, data, retry=None):
...
+
+            if context_response.ok:
+                retry()
+            else:
+                raise Exception('Unable to create missing geoprocessing\n'
```
Comment (rajadain) [inline src/mmw/apps/modeling/geoprocessing.py]: Consider leaving out the `\n` in between the message, so the message is the first line and the details the second. The output should look like:

```
Unable to create missing geoprocessing context in Spark JobServer.
Details = [...]
```

File: src/mmw/apps/modeling/geoprocessing.py [hunk 3]
```diff
@@ -71,6 +72,16 @@ def sjs_submit(host, port, args, data, retry=None):
             except MaxRetriesExceededError:
                 raise Exception('No slots available in Spark JobServer.\n'
                                 'Details = {}'.format(response.text))
+        elif error['result'] == 'context geoprocessing not found':
+            geop_context_url = '{}/contexts?reset=reboot'.format(base_url)
+            context_response = requests.put(geop_context_url)
+
+            if context_response.ok:
+                retry()
```
Comment (rajadain) [inline src/mmw/apps/modeling/geoprocessing.py]: Consider checking that `retry` exists. It defaults to `None`.

**Comment by rajadain** [comment]
I just tried deleting the geoprocessing context and then triggering a MapShed job. I was afraid that rebooting would take too long, and since many calls to `sjs_submit` are made in quick succession (almost in parallel), they would all try to reboot SJS. But in my tests the first one reboots it and the rest pick it up almost immediately, which is great.

I'll see if I can think of any more edge cases, but this is looking really good. Good job!

File: src/mmw/apps/modeling/geoprocessing.py [hunk 4]
```diff
@@ -71,6 +72,21 @@ def sjs_submit(host, port, args, data, retry=None):
...
+
+            if context_response.ok:
+                if retry:
+                    retry()
+                else:
+                    raise Exception('Geoprocessing context missing in Spark'
+                                    'JobServer, but no retry was set.\n'
+                                    'Details = {}'.format())
+            else:
+                raise Exception('Unable to create missing geoprocessing'
```
Comment (rajadain) [inline src/mmw/apps/modeling/geoprocessing.py]: You may need a space at the end of `geoprocessing` else it will be joined with `context` on the next line:

```python
In [1]: print('Unabe to create missing geoprocessing'
   ...: 'context in Spark JobServer.\n')
Unabe to create missing geoprocessingcontext in Spark JobServer.

```

**Comment by rajadain** [comment]
This looks good to me. 👍

File: src/mmw/apps/modeling/geoprocessing.py (hunk 1 above)
Comment (rajadain) [inline src/mmw/apps/modeling/geoprocessing.py]: Consider renaming `geop_context_url` to `reboot_sjs_url` or similar to make it more apparent what the code is trying to do here.

Comment (kellyi) [inline src/mmw/apps/modeling/geoprocessing.py]: Sounds good. I'll make that change to include in the final commit.

**Comment by kellyi** [comment]
This already got put on the `release/1.15.0` branch, then was merged back into `develop` in https://github.com/WikiWatershed/model-my-watershed/commit/90edb189e99848ce262f778996648a212ee603e3

Going to close this PR out.
